stripecardscan/src/main/java/com/stripe/android/stripecardscan/cardscan/CardScanActivity.kt

```diff
@@ -0,0 +1,683 @@
+import com.stripe.android.stripecardscan.cardscan.util.show
+import com.stripe.android.stripecardscan.cardscan.util.startAnimation
+import com.stripe.android.stripecardscan.databinding.ActivityCardscanBinding
+import com.stripe.android.stripecardscan.framework.Config
+import com.stripe.android.stripecardscan.framework.StorageFactory
+import com.stripe.android.stripecardscan.framework.util.getAppPackageName
+import com.stripe.android.stripecardscan.payment.card.ScannedCard
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import kotlinx.parcelize.Parcelize
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.coroutines.CoroutineContext
+import kotlin.math.min
+import kotlin.math.roundToInt
+
+internal const val INTENT_PARAM_REQUEST = "request"
+internal const val INTENT_PARAM_RESULT = "result"
+
+private val MINIMUM_RESOLUTION = Size(1067, 600) // minimum size of OCR
+private const val PERMISSION_RATIONALE_SHOWN = "permission_rationale_shown"
+
+sealed interface CancellationReason : Parcelable {
+
+    @Parcelize
+    object Closed : CancellationReason
+
+    @Parcelize
+    object Back : CancellationReason
+
+    @Parcelize
+    object CameraPermissionDenied : CancellationReason
+}
+
+internal interface ScanResultListener {
+
+    /**
+     * The scan completed.
+     */
+    fun cardScanComplete(pan: String)
+    /**
+     * The user canceled the scan.
+     */
+    fun userCanceled(reason: CancellationReason)
+
+    /**
+     * The scan failed because of an error.
+     */
+    fun failed(cause: Throwable?)
+}
+
+/**
+ * A basic implementation that displays error messages when there is a problem with the camera.
+ */
+internal open class CameraErrorListenerImpl(
+    protected val context: Context,
+    protected val callback: (Throwable?) -> Unit
+) : CameraErrorListener {
+    override fun onCameraOpenError(cause: Throwable?) {
+        showCameraError(R.string.stripe_error_camera_open, cause)
+    }
+
+    override fun onCameraAccessError(cause: Throwable?) {
+        showCameraError(R.string.stripe_error_camera_access, cause)
+    }
+
+    override fun onCameraUnsupportedError(cause: Throwable?) {
+        Log.e(Config.logTag, "Camera not supported", cause)
+        showCameraError(R.string.stripe_error_camera_unsupported, cause)
+    }
+
+    private fun showCameraError(@StringRes message: Int, cause: Throwable?) {
+        AlertDialog.Builder(context)
+            .setTitle(R.string.stripe_error_camera_title)
+            .setMessage(message)
+            .setPositiveButton(R.string.stripe_error_camera_acknowledge_button) { _, _ ->
+                callback(cause)
+            }
+            .show()
+    }
+}
+
+@Keep
+internal open class CardScanActivity : AppCompatActivity(), CoroutineScope {
```

Comment:
this can probably be final instead of open, and we likely don't need the `@Keep` annotation since we don't expect users to be accessing this class.